src/engine/TransitivePathImpl.h

```diff
@@ -235,6 +238,201 @@ class TransitivePathImpl : public TransitivePathBase {
+                               graphId,
+                               std::move(connectedNodes),
+                               mergedVocab_.clone(),
+                               payload,
+                               currentRow};
+        // Reset vocab to prevent merging the same vocab over and over
+        // again.
+        if (yieldOnce_) {
+          mergedVocab_ = LocalVocab{};
+        }
+        timer_.stop();
+        return LoopControl<NodeWithTargets>::yieldValue(std::move(result));
+      }
+      timer_.stop();
+      return LoopControl<NodeWithTargets>::makeContinue();
+    }
+
+    NodeWithTargetRange buildResultRange() {
+      auto& tableColumn = *tableColumnIt_;
+      const auto& [currentRow, zippedType] = *enumerateRangeIt_;
+      return NodeWithTargetRange(
+          ad_utility::CachingContinuableTransformInputRange(
+              tableColumn.expandUndef(zippedType, edges_,
+                                      parent_.graphVariable_.has_value()),
+              [this, currentRow = currentRow,
+               payload = tableColumn.payload_](auto& idPair) mutable {
+                return process(idPair, currentRow, payload);
+              }));
+    }
+
+    void buildEnumerateRange() {
+      if (tableColumnIt_ == ql::ranges::end(startNodes_)) {
+        finished_ = true;
+        return;
+      }
+      auto& tableColumn = *tableColumnIt_;
+      mergedVocab_ = std::move(tableColumn.vocab_);
+      mergedVocab_.mergeWith(edgesVocab_);
+
+      enumerateRange_ = ::ranges::views::enumerate(tableColumn.startNodes_);
+      enumerateRangeIt_ = ql::ranges::begin(*enumerateRange_);
+
+      if (enumerateRangeIt_ == ql::ranges::end(*enumerateRange_)) {
+        // if there are no start nodes in this table column, we move to the
+        // next table column and recursively find the next valid range.
+        ++tableColumnIt_;
+        return buildEnumerateRange();
+      }
+    }
+
+    void getNextResultRange() {
+      ++enumerateRangeIt_;
+      if (enumerateRangeIt_ == ql::ranges::end(*enumerateRange_)) {
+        ++tableColumnIt_;
+        buildEnumerateRange();
+      }
+
+      if (finished_) {
+        return;
+      }
+
+      resultRange_ = buildResultRange();
+      result_ = resultRange_->begin();
+      if (result_ == resultRange_->end()) {
+        getNextResultRange();
+      }
+    }
+
+    void start() {
+      tableColumnIt_ = ql::ranges::begin(startNodes_);
+      buildEnumerateRange();
+      if (finished_) {
+        return;
+      }
+
+      resultRange_ = buildResultRange();
+      result_ = resultRange_->begin();
+      if (result_ == resultRange_->end()) {
+        getNextResultRange();
+      }
+    }
+
+    void next() {
+      ++result_;
+      if (result_ == resultRange_->end()) {
+        getNextResultRange();
+      }
+    }
+
+    bool isFinished() { return finished_; }
+    auto& get() { return *result_; }
+    const auto& get() const { return *result_; }
```

Comment:
Please document all the other newly added functions I didn't explicitly mention. No need to write lengthy comments, a short sentence should suffice.